includes/class-amp-theme-support.php

```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
 				self::redirect_non_amp_url( current_user_can( 'manage_options' ) ? 302 : 301 );
 			}
 
+			if ( MobileRedirectManager::is_enabled() && is_amp_available() ) {
+				$mobile_redirection_disabled = isset( $_GET[ MobileRedirectManager::NO_AMP_QUERY_VAR ] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+				// If `noamp` query var is present, set a cookie to persist disabling of mobile redirection (if one has not already been set).
+				if ( $mobile_redirection_disabled && ! MobileRedirectManager::redirection_disabled() ) {
+					setcookie(
+						MobileRedirectManager::DISABLED_COOKIE_NAME,
+						'1',
+						[
+							'expires'  => 0, // Cookie will only last for the current browser session.
+							'path'     => '/',
+							'secure'   => isset( $_SERVER['HTTPS'] ),
+							'httponly' => false,
+							'samesite' => 'strict',
+						]
+					);
+				}
+
+				// Redirect if `noamp` query parameter is not present and JS redirection is disabled.
+				if ( ! $mobile_redirection_disabled && ! MobileRedirectManager::should_redirect_via_js() ) {
+					$amp_url = add_query_arg( amp_get_slug(), '1', amp_get_current_url() );
+					wp_safe_redirect( $amp_url, current_user_can( 'manage_options' ) ? 302 : 301 );
+				}
+
+				// Add mobile redirection script if user has opted for that solution.
+				if ( MobileRedirectManager::should_redirect_via_js() ) {
+					// The redirect script will add the mobile version switcher link.
+					add_action( 'wp_head', [ MobileRedirectManager::class, 'add_mobile_redirect_script' ], ~PHP_INT_MAX );
+				}
+
+				// Add a link to the footer to allow for navigation to the AMP version.
+				add_action( 'wp_footer', [ __CLASS__, 'add_amp_mobile_version_switcher' ] );
+			}
+
 			amp_add_frontend_actions();
 			return;
 		}
 
 		self::ensure_proper_amp_location();
 
+		if ( ! amp_is_canonical() && MobileRedirectManager::is_enabled() && is_amp_available() ) {
+			// @todo The `amp_post_template_footer` should only be used for the reader mode Classic theme.
```

Comment:
Why the todo?

Reply:
If I understand correctly, the `amp_post_template_footer` action is only applied in the reader mode Classic theme. If that is the case, then we will only need to add to this action if Classic is the selected reader mode theme, otherwise we would use the `wp_footer` action.

Reply:
Right, but isn't that todo already implemented in the ternary below? In any case, it wouldn't hurt to add the `wp_footer` and `amp_post_template_footer` both since the two actions aren't triggered on the same template.

Reply:
Yes, but given the upcoming enhancements to the reader mode, I suppose the conditional below would no longer be correct, given that there will be a variety of themes to choose from for that template mode.

Reply:
Good point. In that case, let's just do two `add_action()` calls here, one for `wp_footer` and another for `amp_post_template_footer`. Only one of them will end up getting used.

Reply:
Sounds good :+1: .